### PR TITLE
Tram integration

### PR DESCRIPTION
**Pull request by nirmal-kumar1**
- [x] Tram infill for MSOAs with tram catchment
- [x] Tram adjustment for North MSOAs
- [x] Tram adjustment for Non-Tram MSOAs

### CODE REVIEW

File: normits_demand/models/notem/tram_inclusion.py [hunk 1]
```diff
@@ -0,0 +1,825 @@
...
+        # Converts DVector to dataframe
+        notem_df = notem_tram_seg.to_df()
+
+        for col, dt in self._target_col_dtypes['notem'].items():
+            notem_df[col] = notem_df[col].astype(dt)
+
+        # Retains only msoa zones that have tram data
+        notem_df = notem_df.loc[notem_df['msoa_zone_id'].isin(tram_data['msoa_zone_id'])]
+        notem_msoa_wo_infill = notem_df.copy()
+        notem_df['c_uniq_id'] = pd_utils.str_join_cols(notem_df, self._join_cols)
```
[the remark this answers is not in the document]
Comment (nirmal-kumar1) [reply normits_demand/models/notem/tram_inclusion.py]: I think this part has got removed after your updates.

File: normits_demand/models/notem/tram_inclusion.py [hunk 2]
```diff
@@ -0,0 +1,825 @@
...
+            Condition for slicing.
+
+        Returns
+        -------
+        notem_sub_df:
+            Returns the subset of the original dataframe.
+
+        """
+        notem_sub_df = notem_df.loc[notem_df['m'] == mode]
+        notem_sub_df = notem_sub_df.sort_values(by='c_uniq_id')
```
[the remark this answers is not in the document]
Comment (nirmal-kumar1) [reply normits_demand/models/notem/tram_inclusion.py]: I think this part has got sorted after your updates.

**Comment by BenTaylor-TfN** [comment]
Updated from master

File: normits_demand/models/notem/tram_inclusion.py [hunk 3]
```diff
@@ -0,0 +1,825 @@
...
+        file_ops.check_file_exists(tram_data)
+        [file_ops.check_file_exists(x) for x in notem_output.values()]
+        file_ops.check_path_exists(export_home)
+
+        # Assign
+        self.tram_data = tram_data
+        self.notem_output = notem_output
+        self.export_home = export_home
+        self.years = list(self.notem_output.keys())
+        self.base_year = min(self.years)
```
[the remark this answers is not in the document]
Comment (nirmal-kumar1) [reply normits_demand/models/notem/tram_inclusion.py]: Removed them

File: normits_demand/models/notem/tram_inclusion.py [hunk 4]
```diff
@@ -0,0 +1,825 @@
...
+
+        # Create a logger
+        logger_name = "%s.%s" % (__name__, self.__class__.__name__)
+        log_file_path = os.path.join(self.export_home, self._log_fname)
+        self._logger = nd.get_logger(
+            logger_name=logger_name,
+            log_file_path=log_file_path,
+            instantiate_msg="Initialised new NoTEM Logger",
+        )
+
```
[the remark this answers is not in the document]
Comment (nirmal-kumar1) [reply normits_demand/models/notem/tram_inclusion.py]: Added a running report.

File: normits_demand/models/notem/tram_inclusion.py [hunk 5]
```diff
@@ -0,0 +1,825 @@
...
+class Tram(NoTEMExportPaths):
+    """
+    Tram Inclusion on NoTEM outputs
+    """
+    # Constants
+    _sort_msoa = ['msoa_zone_id', 'p', 'ca', 'm']
+    _sort_north = ['p', 'ca', 'm']
+    _join_cols = ['msoa_zone_id', 'p', 'ca']
+    base_train = pd.DataFrame()
```
[the remark this answers is not in the document]
Comment (nirmal-kumar1) [reply normits_demand/models/notem/tram_inclusion.py]: Made it an instance variable.

File: normits_demand/models/notem/tram_inclusion.py [hunk 6]
```diff
@@ -0,0 +1,825 @@
...
+            If set to True, it will print out progress updates while
+        running.
+
+        Returns
+        -------
+        tram_data:
+            Returns the tram data as dataframe.
+
+        notem_tram_seg:
+            Returns the notem output dvector in tram segmentation.
```
[the remark this answers is not in the document]
Comment (nirmal-kumar1) [reply normits_demand/models/notem/tram_inclusion.py]: Added segmentation

File: normits_demand/core/definitions/segmentations/hb_p_m7_ca/naming_order.csv [hunk 7]
```diff
@@ -0,0 +1,3 @@
+p
```
[the remark this answers is not in the document]
Comment (nirmal-kumar1) [reply normits_demand/core/definitions/segmentations/hb_p_m7_ca/naming_order.csv:1]: Can be modified to a required segmentation name

File: normits_demand/models/notem/tram_inclusion.py [hunk 8]
```diff
@@ -0,0 +1,825 @@
...
+        )
+        # Adds tram as mode 7
+        cols = list(tram_data)
+        cols.insert(2, 'm')
+        tram_data['m'] = 7
+
+        for col, dt in self._target_col_dtypes['tram'].items():
+            tram_data[col] = tram_data[col].astype(dt)
+
+        tram_data = tram_data.reindex(cols, axis=1)
```
[the remark this answers is not in the document]
Comment (nirmal-kumar1) [reply normits_demand/models/notem/tram_inclusion.py]: Ok. Changed everywhere.

**Comment by BenTaylor-TfN** [comment]
Updated from master

**Review by BenTaylor-TfN — APPROVED** [review]

**Comment by BenTaylor-TfN** [comment]
Happy with the code.

Working on analysing the Tram outputs before fully pulling in

**Pull request merged**